libs/filer/workspace_files_cache.go

```diff
@@ -0,0 +1,425 @@
+	close(e.ch)
+}
+
+// Wait for the operation to complete and return the result.
+func (e *statEntry) wait(ctx context.Context) (fs.FileInfo, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case <-e.ch:
+		return e.info, e.err
+	}
+}
+
+// Mark the stat entry as done.
+func (e *statEntry) markDone(info fs.FileInfo, err error) {
+	e.info = info
+	e.err = err
+	close(e.ch)
+}
+
+// executable is the interface all cacheable calls must implement.
+type executable interface {
+	fmt.Stringer
+
+	execute(ctx context.Context, c *cache)
+}
+
+// cache stores all entries for cacheable Workspace File System calls.
+// We care about caching only [ReadDir] and [Stat] calls.
+type cache struct {
+	f Filer
+	m sync.Mutex
+
+	readDir map[string]*readDirEntry
+	stat    map[string]*statEntry
+
+	// Queue of operations to execute.
+	queue chan executable
+
+	// For tracking the number of active goroutines.
+	wg sync.WaitGroup
+}
+
+func newWorkspaceFilesReadaheadCache(f Filer) *cache {
+	c := &cache{
+		f: f,
+
+		readDir: make(map[string]*readDirEntry),
+		stat:    make(map[string]*statEntry),
+
+		queue: make(chan executable, kMaxQueueSize),
+	}
+
+	ctx := context.Background()
+	for range kNumCacheWorkers {
+		c.wg.Add(1)
+		go c.work(ctx)
+	}
+
+	return c
+}
+
+// work until the queue is closed.
+func (c *cache) work(ctx context.Context) {
+	defer c.wg.Done()
+
+	for {
+		e, ok := <-c.queue
+		if !ok {
+			return
+		}
+		e.execute(ctx, c)
+	}
+}
+
+// enqueue adds an operation to the queue.
+// If the context is canceled, an error is returned.
+// If the queue is full, an error is returned.
+//
+// Its caller is holding the lock so it cannot block.
+func (c *cache) enqueue(ctx context.Context, e executable) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case c.queue <- e:
+		return nil
+	default:
+		return queueFullError{e.String()}
+	}
+}
+
+func (c *cache) completeReadDirForDir(name string, dirEntry fs.DirEntry) {
+	// Add to the stat cache if not already present.
+	if _, ok := c.stat[name]; !ok {
+		e := newStatEntry(name)
+		e.markDone(dirEntry.Info())
+		c.stat[name] = e
+	}
+
+	// Queue a [ReadDir] call for the directory if not already present.
+	if _, ok := c.readDir[name]; !ok {
+		// Create a new cache entry and queue the operation.
+		e := newReadDirEntry(name)
+		err := c.enqueue(context.Background(), e)
+		if err != nil {
+			e.markError(err)
+		}
+
+		// Add the entry to the cache, even if has an error.
+		c.readDir[name] = e
+	}
+}
+
+func (c *cache) completeReadDirForFile(name string, dirEntry fs.DirEntry) {
+	// Skip if this entry is already in the cache.
+	if _, ok := c.stat[name]; ok {
+		return
+	}
+
+	// Create a new cache entry.
+	e := newStatEntry(name)
+
+	// Depending on the object type, we either have to perform a real
+	// stat call, or we can use the [fs.DirEntry] info directly.
+	switch dirEntry.(wsfsDirEntry).ObjectType {
+	case workspace.ObjectTypeNotebook:
```

Comment:
Once the WSFS team supports `repos_export_format` in the list API we can get rid of this. At that point, this cache does not need to be specific for WSFS and could be leveraged for other file systems as well, like UC volumes.

Reply:
Indeed. Added a comment saying the same.